pyccel/parser/semantic.py

```diff
@@ -3682,13 +3709,36 @@ def _visit_ClassDef(self, expr):
         for i in methods:
             self._visit_FunctionDef(i)
 
+        if not any(method.name == '__del__' for method in methods):
+            argument = FunctionDefArgument(Variable(cls.name, 'self', cls_base = cls))
+            scope = self.create_new_function_scope('__del__')
+            del_method = FunctionDef('__del__', [argument], (), [Pass()], cls_name=cls.name, scope=scope)
+            self.exit_function_scope()
+            self.insert_function(del_method)
+            cls.add_new_method(del_method)
+
+        for method in cls.methods:
+            if method.name == '__del__':
+                self._current_function = method.name
+                attribute = [attr for attr in cls.attributes if not attr.on_stack]
+                if attribute:
+                    self._allocs[-1].extend(attribute)
+                    method.body.insert2body(*self._garbage_collector(method.body))
+                    if not self._allocs:
+                        self._allocs.append([])
```

Comment:
I am surprised that this fixes the issue. Can you explain why you think this `if` is necessary.

Or why you made this change:
https://github.com/pyccel/pyccel/pull/1508/commits/94a707d2b7f51287350a7d4b949f7f9c3385be4e